### DISCUSSION TITLE
Feature Suggestion: Base Icon rendering on Dynamic Components (+ possibly a registry)

### DISCUSSION THREAD

**Discussion by jbjhjm** [primefaces/community#1312] [PrimeNG] [answered]
Hi everybody, when updating to primeng 17, I saw that you introduced new icon components and a template-based solution to render different custom icons. With that as a base, I think there is a sleeker approach possible with several added benefits.
Let's start with what the current implementation vs the suggested setup looks like in practice.

## Current setup

in template:
```html
<TimesIcon [styleClass]="'p-dropdown-clear-icon'" (click)="clear($event)" *ngIf="!clearIconTemplate" [attr.data-pc-section]="'clearicon'" />
<span class="p-dropdown-clear-icon" (click)="clear($event)" *ngIf="clearIconTemplate" [attr.data-pc-section]="'clearicon'">
    <ng-template *ngTemplateOutlet="clearIconTemplate"></ng-template>
</span>
```

in component:
```ts
class Dropdown {
  @Input() clearIconTemplate
}
```

To use a custom icon:
```html
<p-dropdown 
  ...
  [clearIconTemplate]="clearIconTemplate"
></p-dropdown>
<ng-template #clearIconTemplate>
... my icon
</ng-template>
```

## Suggested approach - dynamically rendered components

In template:
```html
<!-- dynamicComponent is a custom directive, see "How?" section -->
<ng-container *dynamicComponent="clearIconComponent"></ng-container>
```
in component:
```ts
class Dropdown {
  // is input so that users can pass a custom component, defaults to primeicons component
  // as prime icons are standalone, this way there's not even a need to import them into the component module
  @Input() clearIconComponent = TimesIcon
}
```
To use custom icon:
```html
<p-dropdown [clearIconComponent]="MyIconComponent"></p-dropdown>
```

## one step further - global icon swapping

Instead of remembering to pass correct replacement icons to each and every component instance,
this can free up the way to introduce a simple, global Icon Registry Service which could override an icon everywhere with a simple one-time call.

```ts
AppModule {
  constructor(icons:PrimeIconsRegistry) {
    icons.addOverride(TimesIcon, MyCustomIconComponent).
  }
}
```

## Comparison

Benefits of dynamic components:
- less template code required both for implementation as well as passing a custom icon
- custom icon components are reusable as it is not required to create a custom ng-template everywhere
- optional: global icon overrides via registry

Cons:
- Some users may have to convert their icons to custom components first.

## How?

Since angular 15, ng devs provide an official API to render components on demand. It works with module based components, but really shines with the simplicity of standalone components. Over here we're using dynamic component rendering in production for almost 2 years without any issues.

With support for adding component bindings, even using systems like <fa-icon> is possible.
Could happen like this:
```html
<p-dropdown [clearIconComponent]="{component:FaIconComponent, inputs:{icon:"times"}}"></p-dropdown>
```

If you are interested in using this approach, I can show you how dynamicComponentDirective works and help you with implementing it for the icon system.
Maybe it could even support both Components as well as ng-template references, so that the current icon replacement system would continue to work.

**Comment by cetincakiroglu** [comment, marked as answer]
Hi,

Awesome, I liked that! We'll discuss this. Thanks for the valuable feedback it's good for DX to add this support!
